### PR TITLE
Add dnsapi for Vultr

### PR DESCRIPTION
**Pull request by Fudoshiki**
Fully rewritten using dns_cf api as example

### CODE REVIEW

**Comment by Fudoshiki** [comment]
This completed

**Comment by Neilpang** [comment]
check this guide first: https://github.com/Neilpang/acme.sh/wiki/DNS-API-Dev-Guide

**Comment by Fudoshiki** [comment]
Opened #2374
Documentation updated

File: dnsapi/dns_vultr.sh [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  if ! _get_root "$fulldomain"; then
+    return 1
+  fi
+  _debug _sub_domain "$_sub_domain"
+  _debug _domain "$_domain"
+
+  _debug 'Getting txt records'
+  _vultr_rest GET "dns/records?domain=$_domain"
+
+  if printf "%s\n" "$response" | grep "{\"type\":\"TXT\",\"name\":\"$fulldomain\"" >/dev/null; then
```
Comment (Neilpang) [inline dnsapi/dns_vultr.sh]: do you please not include `{` char ?

use other solutions

File: dnsapi/dns_vultr.sh [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    _err 'Error'
+    return 1
+  fi
+
+  _record="$(echo "$response" | _egrep_o "{[^}]*\"type\"\\s*:\\s*\"TXT\"[^}]*}")"
```
Comment (Neilpang) [inline dnsapi/dns_vultr.sh]: do not use complext  regex expression here.
think about other solutions.

**Comment by Fudoshiki** [comment]
I will rework it

**Comment by Fudoshiki** [comment]
@Neilpang Please review it

File: dnsapi/dns_vultr.sh [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+  _debug 'Getting txt records'
+  _vultr_rest GET "dns/records?domain=$_domain"
+
+  if printf "%s\n" "$response" | grep "\"type\":\"TXT\",\"name\":\"$fulldomain\"" >/dev/null; then
+    _err 'Error'
+    return 1
+  fi
+
+  _record="$(echo "$response" | _egrep_o "\"type\":\"TXT\",\"name\":\"_acme-challenge\",\"data\":\\W*$txtvalue\\W*\",\"priority\":0,\"RECORDID\":\\d*")"
```
Comment (Fudoshiki) [inline dnsapi/dns_vultr.sh]: Don’t know how to replace `\\W*` in **_egrep_o**

Substring from response
```bash
{"type":"TXT","name":"_acme-challenge","data":"\"IikqiCI3suY0u-71M-apJt5Ix-0Ai6qDmmOaN4pkhzs\"","priority":0,"RECORDID":16806343,"ttl":300}
```

This doesn’t work 
```bash
  _record="$(echo "$response" | _egrep_o "\"type\":\"TXT\",\"name\":\"_acme-challenge\",\"data\":\"\\\"$txtvalue\\\"\",\"priority\":0,\"RECORDID\":\\d*")"
```

Comment (Neilpang) [reply dnsapi/dns_vultr.sh]: try this one:

```
_record_id="$( echo "$response" | tr '{}' '\n' | grep '"TXT"'| grep "$txtvalue" | tr ',' '\n' | grep -i 'RECORDID' | cut -d : -f 2)"
```

Comment (Fudoshiki) [reply dnsapi/dns_vultr.sh]: Works fine

**Comment by Neilpang** [comment]
fix the CI errors first:  https://travis-ci.org/Neilpang/acme.sh/jobs/560483218

**Comment by Fudoshiki** [comment]
Do I need fix `dnsapi/dns_namesilo.sh` file too?

**Comment by Neilpang** [comment]
which one ?

**Comment by Fudoshiki** [comment]
https://travis-ci.org/Neilpang/acme.sh/jobs/560494277

**Comment by Neilpang** [comment]
fixed,  it's my fault.

File: dnsapi/dns_vultr.sh [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  _debug 'Getting txt records'
+  _vultr_rest GET "dns/records?domain=$_domain"
+
+  if printf "%s\n" "$response" | grep "\"type\":\"TXT\",\"name\":\"$fulldomain\"" >/dev/null; then
+    _err 'Error'
+    return 1
+  fi
+
+  _record_id="$(echo "$response" | tr '{}' '\n' | grep '"TXT"' | grep "$txtvalue" | tr ',' '\n' | grep -i 'RECORDID' | cut -d : -f 2)"
+
```
Comment (Neilpang) [inline dnsapi/dns_vultr.sh]: please add check for `_record_id` here.

It can be empty.

Comment (Fudoshiki) [reply dnsapi/dns_vultr.sh]: sec

Comment (Fudoshiki) [reply dnsapi/dns_vultr.sh]: Not empty, works fine

Comment (Fudoshiki) [reply dnsapi/dns_vultr.sh]:
```sh
_record_id="$(echo "$response" | tr '{}' '\n' | grep '"TXT"' | grep "$txtvalue" | tr ',' '\n' | grep -i 'RECORDID' | cut -d : -f 2)"
echo "$_record_id"
```

```bash
Removing DNS records.
Removing txt: QDPkHsxNfZJo7L3hQILAP-oNmq0WmQcNWMGtfEh5u4I for domain: _acme-challenge.sharevari.com
16857956
Removed: Success
```

Comment (Fudoshiki) [reply dnsapi/dns_vultr.sh]: txt records added and then removed from my vultr account

Comment (Neilpang) [reply dnsapi/dns_vultr.sh]: I knew it works for you now.
but it never hurts if you check the value.
just check it as what I did for namesilo.

It makes your program more robust.

Comment (Fudoshiki) [reply dnsapi/dns_vultr.sh]: I checked before commiting)

Comment (Fudoshiki) [reply dnsapi/dns_vultr.sh]: Done

**Comment by wintersys-projects** [comment]
Don't know if this is a bug, I am new to acme.sh but when I issue the command

~/.acme.sh/acme.sh --debug --issue --dns dns_vultr -d www.thegalley.uk --server zerossl

and I see "Using Cloudflare doh server" in the debug along with the error message "No Reachable Authority at delegation". As far as I know this request has nothing to do with Cloudflare?

**Pull request merged**